workflow/src/main/java/bio/terra/tanagra/workflow/BuildPathsForHierarchy.java

```diff
@@ -0,0 +1,325 @@
+                          // if there are no relationships to complete the next step, then we've
+                          // reached a root node. just keep the path as is and move on the next one
+                          context.output(KV.of(node, currentPath));
+                        } else {
+                          // if there is a next node, then append it to the path and make it the new
+                          // key
+                          context.output(
+                              KV.of(
+                                  Long.valueOf(nextNodeInPath),
+                                  currentPath + "." + nextNodeInPath));
+                        }
+                      }
+                    }
+                  }));
+    }
+    // example for CoGroupByKey+ParDo iteration above:
+    //    desired result (node,path)
+    //    (a1,"a1.a2.a3")
+    //    (a2,"a2.a3")
+    //    (a3,"a3")
+    //    (b1,"b1.b2")
+    //    (b2,"b2")
+    //    (c1,"c1")
+    //
+    //    childParentKVs (child,parent)
+    //    (a1,"a2")
+    //    (a2,"a3")
+    //    (b1,"b2")
+    //
+    //    [iteration 0] nodePathKVs (next node,path)
+    //    (a1,"a1")
+    //    (a2,"a2")
+    //    (a3,"a3")
+    //    (b1,"b1")
+    //    (b2,"b2")
+    //    (c1,"c1")
+    //    [iteration 0] nodeParentPathJoin (next node,paths,next node parents)
+    //    (a1,["a1"],["a2"])
+    //    (a2,["a2"],["a3"])
+    //    (a3,["a3"],[])
+    //    (b1,["b1"],["b2"])
+    //    (b2,["b2"],[])
+    //    (c1,["c1"],[])
+    //
+    //    [iteration 1] nodePathKVs (next node,path)
+    //    (a2,"a2.a1")
+    //    (a3,"a3.a2")
+    //    (a3,"a3")
+    //    (b2,"b2.b1")
+    //    (b2,"b2")
+    //    (c1,"c1")
+    //    [iteration 1] nodeParentPathJoin (next node,paths,next node parents)
+    //    (a1,[],["a2"])
+    //    (a2,["a2.a1"],["a3"])
+    //    (a3,["a3.a2","a3"],[])
+    //    (b1,[],["b2]")
+    //    (b2,["b2.b1","b2"],[])
+    //    (c1,["c1"],[])
+
+    // swap the key of all pairs from the next node in the path (which should all be root nodes at
+    // this point), to the first node in the path
+    PCollection<KV<Long, String>> nodePathKVs =
+        nextNodePathKVs.apply(
+            ParDo.of(
+                new DoFn<KV<Long, String>, KV<Long, String>>() {
+                  @ProcessElement
+                  public void processElement(ProcessContext context) {
+                    KV<Long, String> kvPair = context.element();
+
+                    String path = kvPair.getValue();
+
+                    // strip out the first node in the path, and make it the key
+                    // e.g. (a3,"a1.a2.a3") => (a1,"a2.a3")
+                    //      (c1,"c1") => (c1,"")
+                    int indexOfFirstPeriod = path.indexOf(".");
+                    String firstNodeInPath;
+                    String pathWithoutFirstNode;
+                    if (indexOfFirstPeriod > 0) {
```

Comment:
opt nit: I'd probably be tempted to split by ".", pop the first element, then join them back together here. A little less dependent on the encoding details